File: experiments/control-run/go/summary.go
package main

import (
	"errors"
	"fmt"
	"math"
	"slices"
	"sort"
	"strings"
	"time"
)

// Slow is one entry of the slowest list.
type Slow struct {
	Duration uint32
	Method   string
	Path     string
	AtText   string
	At       time.Time
	Seq      uint64
}

// Busy is one entry of the busiest list.
type Busy struct {
	Count  uint64
	Method string
	Path   string
}

// Summary is the result of reading every log file.
type Summary struct {
	Top       int
	Requests  uint64
	Errors    uint64
	Successes uint64
	Malformed uint64
	First     time.Time
	Last      time.Time
	Slowest   []Slow
	Busiest   []Busy
}

type busyKey struct{ method, path string }

// Accumulator folds lines into a Summary one at a time, so no file is held
// in memory.
type Accumulator struct {
	sum      Summary
	since    time.Time
	hasSince bool
	seq      uint64
	counts   map[busyKey]uint64
}

// Top bounds.
const (
	MinTop = 1
	MaxTop = 100
)

// ErrTop is returned when top is outside MinTop to MaxTop.
var ErrTop = fmt.Errorf("--top must be %d to %d", MinTop, MaxTop)

// NewAccumulator requires top in 1 to 100. A nil since keeps every line.
func NewAccumulator(top int, since *time.Time) (*Accumulator, error) {
	if top < MinTop || top > MaxTop {
		return nil, ErrTop
	}
	a := &Accumulator{sum: Summary{Top: top}, counts: map[busyKey]uint64{}}
	if since != nil {
		a.since, a.hasSince = *since, true
	}
	return a, nil
}

// AddLine parses one raw line and folds it in.
func (a *Accumulator) AddLine(line string) {
	r, err := ParseLine(line)
	if err != nil {
		a.AddMalformed()
		return
	}
	a.Add(r)
}

// AddMalformed counts a line that could not be parsed.
func (a *Accumulator) AddMalformed() {
	a.sum.Malformed++
}

// Add folds in one record, skipping it if it is before --since.
func (a *Accumulator) Add(r Record) {
	if a.hasSince && r.At.Before(a.since) {
		return
	}
	s := &a.sum
	if s.Requests == 0 || r.At.Before(s.First) {
		s.First = r.At
	}
	if s.Requests == 0 || r.At.After(s.Last) {
		s.Last = r.At
	}
	s.Requests++
	if r.Status >= 500 && r.Status <= 599 {
		s.Errors++
	} else {
		s.Successes++
	}
	path := MaskCards(r.Path)
	a.counts[busyKey{r.Method, path}]++
	a.seq++
	a.insertSlow(Slow{Duration: r.Duration, Method: r.Method, Path: path, AtText: r.AtText, At: r.At, Seq: a.seq})
}

// slowerThan orders by duration descending, timestamp ascending, then input
// order, so the order is total.
func slowerThan(x, y Slow) bool {
	if x.Duration != y.Duration {
		return x.Duration > y.Duration
	}
	if !x.At.Equal(y.At) {
		return x.At.Before(y.At)
	}
	return x.Seq < y.Seq
}

// busierThan orders by count descending, path ascending, then method.
func busierThan(x, y Busy) bool {
	if x.Count != y.Count {
		return x.Count > y.Count
	}
	if x.Path != y.Path {
		return x.Path < y.Path
	}
	return x.Method < y.Method
}

func (a *Accumulator) insertSlow(e Slow) {
	list := a.sum.Slowest
	if len(list) == a.sum.Top && !slowerThan(e, list[len(list)-1]) {
		return
	}
	i := sort.Search(len(list), func(i int) bool { return slowerThan(e, list[i]) })
	list = slices.Insert(list, i, e)
	if len(list) > a.sum.Top {
		list = list[:a.sum.Top]
	}
	a.sum.Slowest = list
}

// Summary returns the finished summary after checking its invariants.
func (a *Accumulator) Summary() (Summary, error) {
	s := a.sum
	s.Slowest = slices.Clone(a.sum.Slowest)
	s.Busiest = make([]Busy, 0, len(a.counts))
	for k, n := range a.counts {
		s.Busiest = append(s.Busiest, Busy{Count: n, Method: k.method, Path: k.path})
	}
	sort.Slice(s.Busiest, func(i, j int) bool { return busierThan(s.Busiest[i], s.Busiest[j]) })
	if len(s.Busiest) > s.Top {
		s.Busiest = s.Busiest[:s.Top]
	}
	if err := s.Check(); err != nil {
		return Summary{}, err
	}
	return s, nil
}

// Invariant violations reported by Check.
var (
	ErrErrorsExceed = errors.New("summary: errors > requests")
	ErrUnbalanced   = errors.New("summary: requests != errors + successes")
	ErrCardLeak     = errors.New("summary: a card number would reach stdout")
	ErrOrder        = errors.New("summary: list out of order or longer than top")
)

// Check enforces the nevers and ordering contracts on a summary.
func (s Summary) Check() error {
	if s.Errors > s.Requests {
		return ErrErrorsExceed
	}
	if s.Requests != s.Errors+s.Successes {
		return ErrUnbalanced
	}
	if len(s.Slowest) > s.Top || len(s.Busiest) > s.Top {
		return ErrOrder
	}
	for i, e := range s.Slowest {
		if HasCard(strings.Join([]string{e.Method, e.Path, e.AtText}, " ")) {
			return ErrCardLeak
		}
		if i > 0 && !slowerThan(s.Slowest[i-1], e) {
			return ErrOrder
		}
	}
	for i, e := range s.Busiest {
		if HasCard(e.Method + " " + e.Path) {
			return ErrCardLeak
		}
		if i > 0 && !busierThan(s.Busiest[i-1], e) {
			return ErrOrder
		}
	}
	return nil
}

// ErrorRate is errors / requests, 0 when there are no requests.
func (s Summary) ErrorRate() float64 {
	if s.Requests == 0 {
		return 0
	}
	return float64(s.Errors) / float64(s.Requests)
}

// PerMinute is requests divided by the first-to-last span in minutes; 0 when
// the span is zero, which covers a single request.
func (s Summary) PerMinute() float64 {
	span := s.Last.Sub(s.First).Minutes()
	if s.Requests < 2 || span <= 0 {
		return 0
	}
	return float64(s.Requests) / span
}

// roundTo rounds x half away from zero to the given number of decimals.
func roundTo(x float64, decimals int) float64 {
	p := math.Pow(10, float64(decimals))
	return math.Round(x*p) / p
}

File: experiments/control-run/python/summary.py
"""Folding records into a Summary, one record at a time."""

from dataclasses import dataclass, field
from datetime import datetime

from parse import Record

TOP_MIN = 1
TOP_MAX = 100


def check_top(top: int) -> None:
    if not TOP_MIN <= top <= TOP_MAX:
        raise ValueError(f"top {top} is outside {TOP_MIN} to {TOP_MAX}")


@dataclass(frozen=True)
class PathCount:
    count: int
    method: str
    path: str

    def __post_init__(self) -> None:
        if self.count < 1:
            raise ValueError(f"count {self.count} must be at least 1")


def slowest_key(record: Record) -> tuple[int, datetime, str, str]:
    return (-record.duration_ms, record.at, record.method, record.path)


def busiest_key(entry: PathCount) -> tuple[int, str, str]:
    return (-entry.count, entry.path, entry.method)


@dataclass(frozen=True)
class Summary:
    top: int
    requests: int
    errors: int
    successes: int
    malformed: int
    first_at: datetime | None
    last_at: datetime | None
    slowest: tuple[Record, ...]
    busiest: tuple[PathCount, ...]

    def __post_init__(self) -> None:
        check_top(self.top)
        if min(self.requests, self.errors, self.successes, self.malformed) < 0:
            raise ValueError("counts must not be negative")
        if self.requests != self.errors + self.successes:
            raise ValueError("requests must equal errors + successes")
        if self.errors > self.requests:
            raise ValueError("errors must not exceed requests")
        self.check_span()
        self.check_lists()

    def check_span(self) -> None:
        if (self.first_at is None) != (self.requests == 0) or (self.last_at is None) != (self.requests == 0):
            raise ValueError("first and last timestamps are set exactly when there are requests")
        if self.first_at is not None and self.last_at is not None and self.first_at > self.last_at:
            raise ValueError("first timestamp must not be after the last")

    def check_lists(self) -> None:
        if len(self.slowest) > min(self.top, self.requests) or len(self.busiest) > min(self.top, self.requests):
            raise ValueError("lists must hold at most top entries and at most one per request")
        if list(self.slowest) != sorted(self.slowest, key=slowest_key):
            raise ValueError("slowest must be sorted by duration descending, then timestamp ascending")
        if list(self.busiest) != sorted(self.busiest, key=busiest_key):
            raise ValueError("busiest must be sorted by count descending, then path ascending")

    @property
    def error_rate(self) -> float:
        return self.errors / self.requests if self.requests else 0.0

    @property
    def per_minute(self) -> float:
        if self.first_at is None or self.last_at is None or self.first_at == self.last_at:
            return 0.0
        minutes = (self.last_at - self.first_at).total_seconds() / 60
        return self.requests / minutes


@dataclass
class Accumulator:
    """Streams records in; keeps counts, the time span, a bounded slowest pool, and per-path counts."""

    top: int
    requests: int = 0
    errors: int = 0
    successes: int = 0
    malformed: int = 0
    first_at: datetime | None = None
    last_at: datetime | None = None
    slowest_pool: list[Record] = field(default_factory=list)
    path_counts: dict[tuple[str, str], int] = field(default_factory=dict)

    def __post_init__(self) -> None:
        check_top(self.top)

    def add_malformed(self) -> None:
        self.malformed += 1

    def add(self, record: Record) -> None:
        self.requests += 1
        if record.is_error:
            self.errors += 1
        else:
            self.successes += 1
        if self.first_at is None or record.at < self.first_at:
            self.first_at = record.at
        if self.last_at is None or record.at > self.last_at:
            self.last_at = record.at
        key = (record.method, record.path)
        self.path_counts[key] = self.path_counts.get(key, 0) + 1
        self.slowest_pool.append(record)
        if len(self.slowest_pool) >= 2 * self.top + 64:
            self.slowest_pool = sorted(self.slowest_pool, key=slowest_key)[: self.top]

    def finish(self) -> Summary:
        busiest = (PathCount(count, method, path) for (method, path), count in self.path_counts.items())
        return Summary(
            top=self.top,
            requests=self.requests,
            errors=self.errors,
            successes=self.successes,
            malformed=self.malformed,
            first_at=self.first_at,
            last_at=self.last_at,
            slowest=tuple(sorted(self.slowest_pool, key=slowest_key)[: self.top]),
            busiest=tuple(sorted(busiest, key=busiest_key)[: self.top]),
        )


def summarize(records: list[Record], top: int, malformed: int = 0) -> Summary:
    accumulator = Accumulator(top)
    for record in records:
        accumulator.add(record)
    for _ in range(malformed):
        accumulator.add_malformed()
    return accumulator.finish()
